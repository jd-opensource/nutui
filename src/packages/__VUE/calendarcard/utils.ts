import Utils from '@/packages/utils/date'
import { CalendarCardDay, CalendarCardMonth, CalendarCardValue } from './types'

export const convertDateToDay = (date: Date) => {
  return date
    ? {
        year: date.getFullYear(),
        month: date.getMonth() + 1,
        date: date.getDate()
      }
    : null
}

export const convertDayToDate = (day: CalendarCardDay) => {
  return day ? new Date(day.year, day.month - 1, day.date) : null
}

export const valueToRange = (val?: CalendarCardValue) => {
  if (Array.isArray(val)) {
    return val.map((date: Date) => {
      return convertDateToDay(date)
    })
  }
  return val ? [convertDateToDay(val)] : []
}

export const rangeTovalue = (range?: CalendarCardDay[]) => {
  if (Array.isArray(range)) {
    return range.map((day: CalendarCardDay) => {
      return convertDayToDate(day)
    })
  }
  return range ? [convertDayToDate(range)] : []
}

/**
 * 获取当月面板中前一个月的日期数据
 */
export const getPrevMonthDays = (year: number, month: number, firstDayOfWeek: number) => {
  let prevMonth = month - 1
  let prevYear = year
  if (prevMonth <= 0) {
    prevMonth = 12
    prevYear -= 1
  }
  let days = Utils.getMonthPreDay(year, month)
  days -= firstDayOfWeek
  if (days >= 7) {
    days -= 7
  }

  const preDates = Utils.getMonthDays(`${prevYear}`, `${prevMonth}`)
  const months = Array.from(Array(preDates), (_, k) => {
    return {
      type: 'prev',
      year: prevYear,
      month: prevMonth,
      date: k + 1
    } as CalendarCardDay
  })
  return months.slice(preDates - days)
}

/**
 * 获取当前月的日期数据
 */
export const getCurrentMonthDays = (year: number, month: number) => {
  const days = Utils.getMonthDays(`${year}`, `${month}`)
  return Array.from(Array(days), (_, k) => {
    return {
      type: 'current',
      year,
      month,
      date: k + 1
    } as CalendarCardDay
  })
}

/**
 * 根据日期获取当前周的起始日期
 */
export const getCurrentWeekDays = (day: CalendarCardDay, firstDayOfWeek: number): CalendarCardDay[] => {
  const current = new Date(day.year, day.month - 1, day.date)
  const count = (current.getDay() + 7 - firstDayOfWeek) % 7
  return [
    convertDateToDay(new Date(current.getTime() - 24 * 60 * 60 * 1000 * count)) as CalendarCardDay,
    convertDateToDay(new Date(current.getTime() + 24 * 60 * 60 * 1000 * (6 - count))) as CalendarCardDay
  ]
}

/**
 * 获取 month 月份数据
 * @param month
 * @param firstDayOfWeek
 * @returns
 */
export const getDays = (month: CalendarCardMonth, firstDayOfWeek: number) => {
  const y = month.year
  const m = month.month
  const days = [...getPrevMonthDays(y, m, firstDayOfWeek), ...getCurrentMonthDays(y, m)] as CalendarCardDay[]
  const size = days.length
  const yearOfNextMonth = month.month === 12 ? month.year + 1 : month.year
  const monthOfNextMonth = month.month === 12 ? 1 : month.month + 1
  // 补全 6 行 7 列视图
  for (let i = 1; i <= 42 - size; i++) {
    days.push({
      type: 'next',
      year: yearOfNextMonth,
      month: monthOfNextMonth,
      date: i
    })
  }
  return days
}

/**
 * 比较两个日期
 * @param day1
 * @param day2
 * @returns
 */
export const compareDay = (day1: CalendarCardDay, day2: CalendarCardDay) => {
  if (day1 && day2) {
    if (day1.year === day2.year) {
      if (day1.month === day2.month) {
        return day1.date - day2.date
      }
      return day1.month - day2.month
    }
    return day1.year - day2.year
  }
}

export const isSameDay = (day1: CalendarCardDay, day2: CalendarCardDay) => {
  return compareDay(day1, day2) === 0
}
