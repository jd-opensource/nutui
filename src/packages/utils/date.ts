const Utils = {
  /**
   * 是否为闫年
   * @return {Boolse} true|false
   */
  isLeapYear: function (y: number): boolean {
    return (y % 4 == 0 && y % 100 != 0) || y % 400 == 0;
  },

  /**
   * 返回星期数
   * @return {String}
   */
  getWhatDay: function (year: number, month: number, day: number): string {
    const date = new Date(year + '/' + month + '/' + day);
    const index = date.getDay();
    const dayNames = ['星期日', '星期一', '星期二', '星期三', '星期四', '星期五', '星期六'];
    return dayNames[index];
  },

  /**
   * 返回星期数
   * @return {Number}
   */
  getMonthPreDay: function (year: number, month: number): number {
    const date = new Date(year + '/' + month + '/01');
    let day = date.getDay();
    if (day == 0) {
      day = 7;
    }
    return day;
  },

  /**
   * 返回月份天数
   * @return {Number}
   */
  getMonthDays: function (year: string, month: string): number {
    if (/^0/.test(month)) {
      month = month.split('')[1];
    }
    return ([0, 31, this.isLeapYear(Number(year)) ? 29 : 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31] as number[])[
      month as any
    ];
  },

  /**
   * 补齐数字位数
   * @return {string}
   */
  getNumTwoBit: function (n: number): string {
    n = Number(n);
    return (n > 9 ? '' : '0') + n;
  },

  /**
   * 日期对象转成字符串
   * @return {string}
   */
  date2Str: function (date: Date, split?: string): string {
    split = split || '-';
    const y = date.getFullYear();
    const m = this.getNumTwoBit(date.getMonth() + 1);
    const d = this.getNumTwoBit(date.getDate());
    return [y, m, d].join(split);
  },

  /**
   * 返回日期格式字符串
   * @param {Number} 0返回今天的日期、1返回明天的日期，2返回后天得日期，依次类推
   * @return {string} '2014-12-31'
   */
  getDay: function (i: number): string {
    i = i || 0;
    let date = new Date();
    const diff = i * (1000 * 60 * 60 * 24);
    date = new Date(date.getTime() + diff);
    return this.date2Str(date);
  },

  /**
   * 时间比较
   * @return {Boolean}
   */
  compareDate: function (date1: string, date2: string): boolean {
    const startTime = new Date(date1.replace('-', '/').replace('-', '/'));
    const endTime = new Date(date2.replace('-', '/').replace('-', '/'));
    if (startTime >= endTime) {
      return false;
    }
    return true;
  },

  /**
   * 时间是否相等
   * @return {Boolean}
   */
  isEqual: function (date1: string, date2: string): boolean {
    const startTime = new Date(date1).getTime();
    const endTime = new Date(date2).getTime();
    if (startTime == endTime) {
      return true;
    }
    return false;
  },
  getMonthWeek: function (year: string, month: string, date: string, firstDayOfWeek: number = 0): number {
    const dateNow = new Date(Number(year), parseInt(month) - 1, Number(date));
    let w = dateNow.getDay(); //星期数
    let d = dateNow.getDate();
    let remainder = 6 - w;
    if (firstDayOfWeek !== 0) {
      w = w == 0 ? 7 : w;
      remainder = 7 - w;
    }
    return Math.ceil((d + remainder) / 7);
  },
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  getYearWeek: function (year: string, month: string, date: string, firstDayOfWeek: number = 0): number {
    const dateNow = new Date(Number(year), parseInt(month) - 1, Number(date));
    const dateFirst = new Date(Number(year), 0, 1);
    const dataNumber = Math.round((dateNow.valueOf() - dateFirst.valueOf()) / 86400000);
    return Math.ceil((dataNumber + (dateFirst.getDay() + 1 - 1)) / 7);
  },
  getWeekDate: function (year: string, month: string, date: string, firstDayOfWeek: number = 0): string[] {
    const dateNow = new Date(Number(year), parseInt(month) - 1, Number(date));
    const nowTime = dateNow.getTime();
    let day = dateNow.getDay();
    if (firstDayOfWeek === 0) {
      const oneDayTime = 24 * 60 * 60 * 1000;
      //显示周日
      const SundayTime = nowTime - day * oneDayTime; //本周的周日
      //显示周六
      const SaturdayTime = nowTime + (6 - day) * oneDayTime; //本周的周六

      const sunday = this.date2Str(new Date(SundayTime));
      const saturday = this.date2Str(new Date(SaturdayTime));
      return [sunday, saturday];
    } else {
      day = day == 0 ? 7 : day;
      const oneDayTime = 24 * 60 * 60 * 1000;
      //显示周一
      const MondayTime = nowTime - (day - 1) * oneDayTime; //本周的周一
      //显示周日
      const SundayTime = nowTime + (7 - day) * oneDayTime; //本周的周日

      const monday = this.date2Str(new Date(MondayTime));
      const sunday = this.date2Str(new Date(SundayTime));
      return [monday, sunday];
    }
  },
  formatResultDate: function (date: string) {
    let days = [...date.split('-')];
    days[2] = Utils.getNumTwoBit(Number(days[2]));
    days[3] = `${days[0]}-${days[1]}-${days[2]}`;
    days[4] = Utils.getWhatDay(+days[0], +days[1], +days[2]);
    return days;
  }
};

export default Utils;
