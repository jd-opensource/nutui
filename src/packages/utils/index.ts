export * from './useRelation';
export * from './useLocale';
