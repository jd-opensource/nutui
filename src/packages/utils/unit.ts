import { ComponentPublicInstance, nextTick } from 'vue';
import { VueWrapper, DOMWrapper } from '@vue/test-utils';

function getTouch(el: Element | Window, x: number, y: number) {
  return {
    identifier: Date.now(),
    target: el,
    pageX: x,
    pageY: y,
    clientX: x,
    clientY: y,
    radiusX: 2.5,
    radiusY: 2.5,
    rotationAngle: 10,
    force: 0.5
  };
}

export function trigger(
  wrapper: VueWrapper<ComponentPublicInstance<any, any, any>> | DOMWrapper<Element> | Element | Window,
  eventName: string,
  x = 0,
  y = 0,
  options: any = {}
) {
  const el = 'element' in wrapper ? wrapper.element : wrapper;
  const touchList = options.touchList || [getTouch(el, x, y)];

  if (options.x || options.y) {
    touchList.push(getTouch(el, options.x, options.y));
  }

  const event = document.createEvent('CustomEvent');
  event.initCustomEvent(eventName, true, true, {});

  Object.assign(event, {
    clientX: x,
    clientY: y,
    touches: touchList,
    targetTouches: touchList,
    changedTouches: touchList
  });

  el.dispatchEvent(event);

  return nextTick();
}
export async function mockScrollTop(value: number) {
  Object.defineProperty(window, 'scrollTop', { value, writable: true });
  trigger(window, 'scroll');
  return nextTick();
}

export function sleep(delay = 0): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(resolve, delay);
  });
}
