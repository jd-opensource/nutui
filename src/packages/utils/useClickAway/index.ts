import { Ref, isRef, onMounted, onUnmounted } from 'vue';

type TargetType = Element | Ref<Element | undefined>;

export function useClickAway(onClickAway: () => void, target: TargetType | TargetType[], eventName = 'touchstart') {
  const handler = (event: Event) => {
    const targets = Array.isArray(target) ? target : [target];
    const inTarget = targets.some((item) => {
      const targetElement = isRef(item) ? item.value : item;
      return !targetElement || targetElement.contains(event.target as Node);
    });
    if (!inTarget) {
      onClickAway();
    }
  };

  onMounted(() => {
    window.addEventListener(eventName, handler, true);
  });

  onUnmounted(() => {
    window.removeEventListener(eventName, handler, true);
  });
}
