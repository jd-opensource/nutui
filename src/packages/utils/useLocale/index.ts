import locale from '@/packages/locale';
import { getPropByPath, isFunction } from '../util';

export const useLocale = (name = '') => {
  return (keyPath: string, ...args: unknown[]) => {
    name = name.toLocaleLowerCase();
    const languages = locale.languages();
    let text = keyPath;
    if (name && name.startsWith('nut')) {
      text = `${name.slice(3)}.${keyPath}`;
    }
    const res = getPropByPath(languages, text) || getPropByPath(languages, keyPath);
    return isFunction(res) ? res(...args) : res;
  };
};
