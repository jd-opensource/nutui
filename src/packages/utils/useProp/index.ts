import { computed, getCurrentInstance } from 'vue';
import type { ComponentPublicInstance, ComputedRef } from 'vue';

export const useProp = <T>(name: string): ComputedRef<T | undefined> => {
  const vm = getCurrentInstance();
  if (!vm) {
    throw new Error('useProp must be called within a setup function');
  }
  const proxy = vm.proxy as ComponentPublicInstance;
  const props = proxy?.$props as Record<string, T>;
  return computed(() => props?.[name]);
};
