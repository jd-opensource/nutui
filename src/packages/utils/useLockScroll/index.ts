import { body } from '@/packages/utils/util'
let count = 0

const CLSNAME = 'nut-overflow-hidden'

export const useLockScroll = (isLock: () => boolean) => {
  const lock = () => {
    if (isLock()) {
      try {
        !count && body.classList.add(CLSNAME)
        count++
      } catch (error) {
        console.warn('[NutUI] <useLockScroll>', error)
      }
    }
  }

  const unlock = () => {
    if (isLock() && count) {
      try {
        count--
        !count && body.classList.remove(CLSNAME)
      } catch (error) {
        console.warn('[NutUI] <unlock>', error)
      }
    }
  }

  return [lock, unlock]
}
