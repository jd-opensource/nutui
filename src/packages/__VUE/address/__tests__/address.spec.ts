import { mount } from '@vue/test-utils';
import { nextTick } from 'vue';
import Address from '../index.vue';
import { addressListData, addressExistData } from '../address-list';
import { mockScrollTo } from '@/packages/utils/test';

test('address render', async () => {
  const wrapper = mount(Address, {
    props: {
      visible: true,
      province: addressListData.province,
      city: addressListData.city,
      country: addressListData.country,
      town: addressListData.town,
      teleportDisable: false
    }
  });
  await nextTick();
  expect(wrapper.find('.nut-address__header-close').exists()).toBeTruthy();
  expect(wrapper.findAll('.nut-address__detail-item').length).toBe(5);
});

test('choose address item', async () => {
  const wrapper = mount(Address, {
    props: {
      visible: true,
      province: addressListData.province,
      city: addressListData.city,
      country: addressListData.country,
      town: addressListData.town,
      teleportDisable: false
    }
  });
  const fn = mockScrollTo();
  wrapper.find('.nut-address__detail-item').trigger('click');
  await nextTick();
  expect(fn).toHaveBeenCalledTimes(1);
  expect(wrapper.find('.nut-address__region-item ').text()).toEqual('北京');
});

test('default choose address', async () => {
  const wrapper = mount(Address, {
    props: {
      modelValue: [1, 7, 3],
      visible: true,
      province: addressListData.province,
      city: addressListData.city,
      country: addressListData.country,
      town: addressListData.town,
      teleportDisable: false
    }
  });
  wrapper.vm.initCustomSelected();
  await nextTick();
  expect(wrapper.findAll('.nut-address__region-item ')[0].text()).toEqual('北京');
  expect(wrapper.findAll('.nut-address__region-item ')[1].text()).toEqual('朝阳区');
  expect(wrapper.findAll('.nut-address__region-item ')[2].text()).toEqual('八里庄街道');
});

test('Exist address', async () => {
  const wrapper = mount(Address, {
    props: {
      type: 'exist',
      visible: true,
      existAddress: addressExistData,
      isShowCustomAddress: false,
      teleportDisable: false
    }
  });
  await nextTick();
  expect(wrapper.find('.choose-other').exists()).toBeFalsy();
  expect(wrapper.findAll('.nut-address__exist-group-item').length).toBe(3);
});

test('Exist address choose event', async () => {
  const wrapper = mount(Address, {
    props: {
      type: 'exist',
      visible: true,
      existAddress: addressExistData,
      teleportDisable: false
    }
  });
  await nextTick();
  wrapper.find('.nut-address__exist-group-item').trigger('click');
  const chooseAddress = wrapper.emitted().selected[0];
  expect((chooseAddress as []).length).toBe(3);
});

test('Exist address & list address', async () => {
  const wrapper = mount(Address, {
    props: {
      type: 'exist',
      visible: true,
      existAddress: addressExistData,
      teleportDisable: false,
      province: addressListData.province,
      city: addressListData.city,
      country: addressListData.country,
      town: addressListData.town
    }
  });
  await nextTick();
  const changeBtn = wrapper.find('.nut-address__exist-choose-btn');
  expect(changeBtn.exists()).toBeTruthy();

  changeBtn.trigger('click');
  await nextTick();
  expect(wrapper.find('.nut-address__header-back').exists()).toBeTruthy();
  expect(wrapper.find('.nut-address__custom').exists()).toBeTruthy();
});
