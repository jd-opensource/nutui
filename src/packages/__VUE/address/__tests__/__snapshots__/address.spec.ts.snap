// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Address: choose address item 1`] = `
"<teleport-stub to="body" disabled="true">
  <transition-stub name="overlay-fade" appear="false" persisted="true" css="true">
    <view class="nut-overlay" style="transition-duration: 0.3s; z-index: 2000;"></view>
  </transition-stub>
  <transition-stub name="nut-popup-slide-bottom" appear="false" persisted="true" css="true">
    <view class="nut-popup nut-popup--bottom" style="z-index: 2000; transition-duration: 0.3s;">
      <view class="nut-address">
        <view class="nut-address__header">
          <view class="nut-address__header-back"><svg class="nut-icon nut-icon-left" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1024 1024" role="presentation" size="14px" style="display: none;">
              <path d="M642.973 1005.542 321.912 587.098a123.278 123.278 0 0 1 0-150.17l321.06-418.443a47.182 47.182 0 0 1 74.83 57.422l-321.088 418.47a28.994 28.994 0 0 0 0 35.3l321.088 418.47a47.155 47.155 0 0 1-74.83 57.395" fill="currentColor" fill-opacity="0.9"></path>
            </svg></view>
          <view class="nut-address__header__title">请选择所在地区</view>
          <view class="nut-address__header-close"><svg class="nut-icon nut-icon-close" style="color: rgb(204, 204, 204);" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1026 1024" role="presentation" size="14px">
              <path d="M981.577 1024c-11.703 0-23.406-2.926-32.183-11.703L13.166 76.07c-14.629-17.555-14.629-46.812 0-64.366 17.554-14.629 46.811-14.629 64.365 0L1013.76 947.93c17.554 17.555 17.554 43.886 0 61.44-8.777 11.703-20.48 14.629-32.183 14.629zm-936.228 0c-11.703 0-23.406-2.926-32.183-11.703-17.555-17.554-17.555-43.886 0-61.44L949.394 14.63c17.555-17.555 43.886-17.555 61.44 0 17.555 17.554 17.555 43.885 0 61.44L74.606 1012.297C68.754 1021.074 57.05 1024 45.349 1024z" fill="currentColor" fill-opacity="0.9"></path>
            </svg></view>
        </view><!-- 请选择 -->
        <view class="nut-address__custom">
          <view class="nut-address__region">
            <view class="active nut-address__region-item">
              <view>请选择</view>
            </view>
            <view class="nut-address__region-line" style="left: 20px;"></view>
          </view>
          <view class="nut-address__detail">
            <ul class="nut-address__detail-list">
              <li class="nut-address__detail-item">
                <div>
                  <!--v-if-->北京
                </div>
              </li>
              <li class="nut-address__detail-item">
                <div>
                  <!--v-if-->广西
                </div>
              </li>
              <li class="nut-address__detail-item">
                <div>
                  <!--v-if-->江西
                </div>
              </li>
              <li class="nut-address__detail-item">
                <div>
                  <!--v-if-->四川
                </div>
              </li>
              <li class="nut-address__detail-item">
                <div>
                  <!--v-if-->浙江
                </div>
              </li>
            </ul>
          </view>
        </view>
      </view>
      <!--v-if-->
    </view>
  </transition-stub>
</teleport-stub>"
`;
